test/unittests/execute_floating_point_test.cpp

```diff
@@ -700,6 +700,89 @@ TEST(execute_floating_point, f64_promote_f32)
     EXPECT_GT(FP{res4.value.f64}.nan_payload(), FP64::canon);
 }
 
+TEST(execute_floating_point, f32_demote_f64)
+{
+    /* wat2wasm
+    (func (param f64) (result f32)
+      local.get 0
+      f32.demote_f64
+    )
+    */
+    const auto wasm = from_hex("0061736d0100000001060160017c017d030201000a070105002000b60b");
+    auto instance = instantiate(parse(wasm));
+
+    constexpr double f32_max = FP32::Limits::max();
+    ASSERT_EQ(f32_max, 0x1.fffffep127);
+
+    // The "artificial" f32 range limit: the next f32 number that could be represented
+    // if the exponent had a larger range.
+    // Wasm spec Rounding section denotes this as the limit_N in the float_N function (for N=32).
+    // https://webassembly.github.io/spec/core/exec/numerics.html#rounding
+    constexpr double f32_limit = 0x1p128;  // 2**128.
+
+    // The lower boundary input value that results in the infinity. The number is midway between
+    // f32_max and f32_limit. For this value rounding prefers infinity, because f32_limit is even.
```

Comment:
I don't quite understand the last sentence. `f32_limit` would indeed be even if casted to int, but for floats there's no even/odd ?

Reply:
Ah I guess it's about evenness of the mantissa part

![image](https://user-images.githubusercontent.com/1863135/90234072-5767e300-de1f-11ea-8845-72f5f36c603d.png)

Reply:
Of course there is (at least in Wasm spec).
![image](https://user-images.githubusercontent.com/573380/90234446-f8569e00-de1f-11ea-805d-eaee5bdbf599.png)
In this case `f32_limit` is even by definition.

Reply:
But still confused, mantissa of `f32_max` is even, manitssa of `f32_limit` is odd

Reply:
I think in both cases you are confused by the size of the f32 mantisa which is 23 bits. When written as 6 hex digits the last bit should always be 0. The mantissa full of ones is `1.fffffe` (leading 1 is implicit). Then `m = 0x7fffff` and is odd.

The `f32_limit` is even by definition: `even_N(+-limit_N) <=> true`. But also mantissa of `1p128` is all zero therefore even.

Reply:
I see, yes, I was confused by hex notation.